### PR TITLE
Constructor error message

### PR DESCRIPTION
**Pull request by deanbrophy**
Addresses #11381.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 1]
```diff
@@ -201,12 +201,22 @@ module.exports = function(
...
       const name = getComponentName(workInProgress);
       const renderPresent = instance.render;
-      warning(
-        renderPresent,
-        '%s(...): No `render` method found on the returned component ' +
-          'instance: you may have forgotten to define `render`.',
-        name,
-      );
+
+      if (type.prototype && type.prototype.render !== undefined) {
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: Let's put the whole thing into `if (!renderPresent)` so we don't run those checks unnecessarily. And make it `warning(false`.

File: packages/react-reconciler/src/ReactFiberClassComponent.js [hunk 2]
```diff
@@ -202,21 +202,24 @@ module.exports = function(
...
-          '%s(...): No `render` method found on the returned component ' +
-            'instance: you may have forgotten to define `render`.',
-          name,
-        );
+      if (!renderPresent) {
+        if (type.prototype && type.prototype.render !== undefined) {
+          warning(
+            false,
+            '%s(...): No `render` method found on the returned component ' +
+              'instance: is the `constructor` defined well?',
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberClassComponent.js]: Let's be explicit: "did you accidentally return an object from the constructor"

**Comment by gaearon** [comment]
Do you mind writing a test inside `ReactCompositeComponent-test` against `react-dom` instead?
That's where such tests currently are, for better or worse.

**Comment by deanbrophy** [comment]
Sure thing! I'll have an updated PR over later today.

File: packages/react-dom/src/__tests__/ReactCompositeComponent-test.js [hunk 3]
```diff
@@ -1500,4 +1500,45 @@ describe('ReactCompositeComponent', () => {
...
+      render() {
+        return <div />;
+      }
+    }
+
+    expect(function() {
+      ReactTestUtils.renderIntoDocument(<RenderTextInvalidConstructor />);
+    }).toThrow();
+
+    const error = console.error.calls.mostRecent().args[0];
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactCompositeComponent-test.js]: Let's also assert we only emit one warning. Like other tests do by asserting on count.

File: packages/react-dom/src/__tests__/ReactCompositeComponent-test.js [hunk 4]
```diff
@@ -1500,4 +1500,45 @@ describe('ReactCompositeComponent', () => {
...
+
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactCompositeComponent-test.js]: Nit: let's remove this newline

File: packages/react-dom/src/__tests__/ReactCompositeComponent-test.js [hunk 5]
```diff
@@ -1500,4 +1500,45 @@ describe('ReactCompositeComponent', () => {
...
+  it('should return error if render is not defined', () => {
+    spyOn(console, 'error');
+    class RenderTestUndefinedRender extends React.Component {}
+
+    expect(function() {
+      ReactTestUtils.renderIntoDocument(<RenderTestUndefinedRender />);
+    }).toThrow();
+
+    const error = console.error.calls.mostRecent().args[0];
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactCompositeComponent-test.js]: Same feedback

**Comment by gaearon** [comment]
Looks great. Thanks!

**Comment by deanbrophy** [comment]
@gaearon No problem! Do you have any recommendations for my next issue? Maybe something a little more difficult?

**Comment by gaearon** [comment]
Adding Flow coverage to files that don't currently have it (and fixing Flow errors) would be greatly appreciated.

**Comment by deanbrophy** [comment]
Yep, I'm game for adding flow coverage. Is there a list I can work from? Also, would you want a PR for each file?

**Comment by gaearon** [comment]
No particular list, but you can go through packages and do this in the order you see files without `@flow`. PR per a few related files (especially if they import each other) would probably be best. And it's better to start with untyped files that are imported *from* already typed files. So that it's expanding one "island" of coverage instead of creating multiple independent "islands" that might conflict with each other at seams later.

**Comment by deanbrophy** [comment]
Understood! I'll get a PR over with the first set by the weekend, maybe sooner.

**Comment by deanbrophy** [comment]
@gaearon,
I created PR 11465 which includes Flow types for EventPluginHub. Let me know if I'm heading in the right direction. Also, I opened a Flow issue, [5320](https://github.com/facebook/flow/issues/5320). I ran into a type problem that I resolved by reordering the union type.

**Comment by gaearon** [comment]
There's a few older PRs that I haven't gotten to reviewing yet, so this will take a little more time. I'm also on vacation next week. I'll likely pick it up after that!

**Pull request merged**